Add Batch Importing for .prices files into Trade Dangerous

Ryanel opened the pull request:
Hello!

While flying around the galaxy, I use Elite Dangerous Market Connector to get Trade Dangerous prices files. I import these files so that I can get more accurate advice from Trade Dangerous without having to download from eddb, which takes about 25 minutes for me.

I end up with a lot of .prices files, so I wrote a tiny script to import them all into Trade Dangerous. However, that took way too long to import, up to 10 minutes with 15 or so files! This is because I had to import the files one by one.

The example on the wiki, where you type `trade.py import *.prices` error-ed out on Windows 10. It doesn't look like the code handles multiple files. 

So, I wrote this batch importer to speed up the process. It simply delays regenerating the prices file until all files have been processed.

The way you invoke it is: `trade.py import --batch "file1;file2;file3"`, with paths separated by a semicolon. The quotes are optional, but necessary if any path has a space.

Using this, my import time has gone down to under a minute for 25~ files.

I'd like a quick look over on this @eyeonus , since this is my first time touching your code-base, and I don't want to do something like corrupting the database.

eyeonus commented:
I don't like the idea of adding it directly to the code.

If you reformat this to be a plugin, that would be splendid.

eyeonus commented:
On that note, simply concatenating the files together at run time would also do the trick, so you could literally write a plugin, with no options, that simply takes all the filepaths passed to it and combines them into one file, sets "cmdenv.filename" to be that new file, and returns true to let TD do the processing.

Ryanel commented:
Huh, I'm pretty unfamiliar with the format / plugin, so that sounds good. I can definitely do that.

Ryanel commented:
Hm, I've run into issues using the unmodified import command code. It won't allow for variable amounts of arguments, thinking that anything is a filename (which isn't allowed with --plug); this happens even when accessing argv directly. Using plugin options doesn't work either. The only way I can think of doing this would be to keep the "file1;file2" convention and make --plug and filename switches non mutually exclusive. That way I could access filename from the plugin. Is that acceptable, @eyeonus ?

eyeonus commented:
Ah, I see, you have to have an option. Whoops.

So, create a single option, called something like "files".

That'll allow you to do, for example, 'trade.py import -P batch_prices -O files="file1.prices;file2.prices;..."

See, for example, how it's done with the "test" option on the edapi plugin:
https://github.com/eyeonus/Trade-Dangerous/blob/f6e88f044c10d12781850a0fe0be164c36993ce5/plugins/edapi_plug.py#L439-L452
https://github.com/eyeonus/Trade-Dangerous/blob/f6e88f044c10d12781850a0fe0be164c36993ce5/plugins/edapi_plug.py#L702-L715

eyeonus commented:
Quick modification to make the edapi plugin print what the "test" option is equal to when passing "test=X":

```trade.py import -P edapi -O test="filename;jon;space test"
filename;jon;space test```

Can not use commas, and the entire argument must be enclosed in " if anything has a space, but you can definitely use that to do what you want.

So, you'd need to take the "files" argument, do a split on ";", which will give you the entire file list as individual entries in a list, open a file for writing, and then open and read each file in the list, writing them into the opened file, close the file, set cmdenv.filename to be the file you created, and then return from the plugin passing true.

eyeonus commented:
Cool beans. Looking forward to seeing it.

Ryanel commented:
Hmm, when importing the same station twice, I get an error of `trade.py: tmp\batch_prices.prices:302 ERROR Second occurrance of station "ERAVATE/CLEVE HUB", previous entry at line 7.`

This is more complicated now. I need to somehow only include the latest data from each station, once. This is gonna require some parsing, and more work.

Other then that, the file option and folder option DO work. You can even combine them, such as -Ofolder="path",file="file1;file2;file3"

Ryanel commented:
Or rather, there needs to be some way to determine which file contains the latest data.

eyeonus commented:
You can use the modified time to figure out which file was created more recently. Give me a sec and I'll link in how I did that in the eddblink plugin.

eyeonus commented:
It does sound like, except for the duplication issue, you've got it all figured out.

Ryanel commented:
I believe that issue is fixed. Didn't have to do much parsing. I used the stat() like you suggested. Tell me how this looks!

eyeonus commented:
Looks good to me.

Just to verify, have you tested it, does it work, and what tests have you done?

Off the top of my head, these are the tests I can think of:
Single ,prices file.
Two .prices files.
Two .prices files with the same station.
Folder with multiple .prices files.
Folder with multiple .prices files, many of which have repeating stations.
Single .txt file.
Folder with at least one .prices file and at least one file that is not .prices

All of the above would be in the usage case that each file has one and only one station in it.
Checking against duplicates when there are more than one station in a .prices file would be a lot more complicated, but we don't need to worry about that because: this plugin is built with the intention of using the EDMC generated files, which only have one station per file, and if the .prices file had more than one station, it would probably have a __lot__ more than one, and so it would smarter to use the regular import, not this plugin.

That leads to a thought, actually. Since it is intended specifically to import EDMC's .prices files as a batch, it would probably be a good idea to change the plugin name to reflect that, and it might be a good idea to put in the default folder EDMC stores the files it generates as the default, no options passed, method of running.

I'll give you time to consider, test those cases you haven't if any, and get back to me.

If you say it's good to go as is, taking the above into consideration, I'll merge it into TD once I get your say so.

Ryanel commented:
Hi!

# Test Results
Single .prices file: Works!
Two .prices files. Works!
Two .prices files with the same station : Works, selects newer one.
Folder with multiple .prices files: Works
Folder with multiple .prices files, many of which have repeating stations: Works, select newer per station.
Single .txt file : Works with -Ofiles, does not work with -Ofolder. 
Folder with at least one .prices file and at least one file that is not .prices: Only imports .prices files.
File with invalid contents: Throws an error.
File that's blank: Does not effect output at all. Is not included in output.
File with multiple stations: Throws a plugin error


The de-duplication should be good no matter what's thrown at it. I do that as a last step, given a list of file paths. For folders, I glob *.prices files. Since we're not supporting more then one station per file, I've added an error check. I've also renamed it edmc_batch

I don't like the idea of finding the default folder in the plugin because 
1. It requires code for each supported OS
2. Requires mimicking EDMC's behavior, which may change.

I feel like this is good as is. If someone else wanted to add that behavior, they could. I feel like it's a solid plugin as is.

bgol commented:
It's technicaly an EDDB bug because it's not using the ingame names like EDMC (and the old TD) does. The current TD just inherited the bug from EDDB because it's now using EDDB data.
It should be an easy fix to make a mapping table of the diffs between the EDDB commodities.json and EDMC commodity.csv.

eyeonus commented:
Should be a simple matter to integrate that code into the edmc_batch plugin.

Wasn't aware that EDDB was actually the one in the wrong. Be nice if there was a way to submit a bug report to them and get that fixed. Oh well.

On that, I do know that TD does actually call it, 'Salvageable Wreckage'.
It might actually be a good idea to actually fix things the other way around and make it so that TD corrects the incorrect names EDDB is using.

I might get to it at some point in the future when I'm not overworked to the point of exhaustion.

Tromador commented:
Getting a message to themroc about an eddb bug wouldn't be hard, just informal. 

Having him decide it is a bug and that it's important enough or even desireable to fix maybe less so. We are working around the issue because TD predates it (we date back to beta, prebeta? before there was an EDDB even, certainly), but I suspect a lot of newer tools have been written from the ground up to expect exactly what EDDB puts out, so I further suspect it's been that way for so long now that fixing it would actually break a lot of things which are expecting the "buggy" output.  

Accordingly, I think were I themroc, I might well say "working as designed".

eyeonus commented:
Yeah, that's about what I'd guess would happen as well.

eyeonus commented:
So, the only thing that would have a problem with the item names being changed is the listener, which means that any update to TD needs an equivalent update in the listener as well, which isn't a horrible deal.

I've gone ahead and updated both to start using the correct names for those items.
